File: src/main/java/org/example/kalkulationsprogramm/dto/Freigabe/FreigabeAnsichtDto.java
package org.example.kalkulationsprogramm.dto.Freigabe;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

/**
 * Daten, die Astro auf der öffentlichen Freigabe-Seite anzeigen darf.
 * Bewusst keine internen IDs oder Hashwerte – nur, was der Kunde sehen soll.
 */
@Getter
@Setter
@Builder
@AllArgsConstructor
public class FreigabeAnsichtDto
{
    private String uuid;
    private String status;
    private String dokumentNummer;
    private String dokumentArt;
    private BigDecimal dokumentBetrag;
    private String bauvorhaben;
    private String kundeName;
    private String kundeEmail;
    private LocalDateTime erstelltAm;
    private LocalDateTime ablaufDatum;
    private LocalDateTime akzeptiertAm;
    private boolean abgelaufen;
    /** Pfad-Suffix, über das Astro die PDF beim ERP nachladen kann. */
    private String pdfPfad;

    /**
     * Positionen des Dokuments (HTML-Ansicht). {@code null}, wenn das Dokument keine
     * Positionen führt (z.B. Alt-Freigaben des Anfrage-/Projekt-Dokumentsystems).
     */
    private List<FreigabePositionDto> positionen;

    /** Summe aller NICHT-optionalen Leistungen (netto) – Basis ohne Alternativen. */
    private BigDecimal basisNetto;

    /** Basis-Betrag brutto (ohne Alternativen). */
    private BigDecimal basisBrutto;

    /** MwSt-Satz in Prozent (z.B. 19.0) für die Live-Berechnung der Endsumme im Frontend. */
    private BigDecimal mwstProzent;

    /** true, wenn das Dokument wählbare Alternativpositionen enthält. */
    private boolean hatAlternativen;

    /**
     * Dokument-Pauschalrabatt in Prozent (z.B. 3.0) oder {@code null}, wenn keiner
     * gewährt wurde.
     *
     * <p>Ohne dieses Feld ergibt die Summe der einzeln ausgewiesenen Positionen auf
     * der Freigabe-Seite mehr als der angezeigte Gesamtbetrag — der Kunde sähe die
     * Differenz ohne Erklärung. {@code basisNetto}/{@code basisBrutto} sind bereits
     * die rabattierten Werte; dieses Feld dient allein dem Ausweis der Zeile
     * "Rabatt −x %".</p>
     */
    private BigDecimal globalRabattProzent;
}

File: src/main/kotlin/org/example/kalkulationsprogramm/dto/Freigabe/FreigabeAnsichtDto.kt
package org.example.kalkulationsprogramm.dto.Freigabe

import java.math.BigDecimal
import java.time.LocalDateTime

class FreigabeAnsichtDto(
    var uuid: String?,
    var status: String?,
    var dokumentNummer: String?,
    var dokumentArt: String?,
    var dokumentBetrag: BigDecimal?,
    var bauvorhaben: String?,
    var kundeName: String?,
    var kundeEmail: String?,
    var erstelltAm: LocalDateTime?,
    var ablaufDatum: LocalDateTime?,
    var akzeptiertAm: LocalDateTime?,
    var abgelaufen: Boolean,
    var pdfPfad: String?,
    var positionen: List<FreigabePositionDto>?,
    var basisNetto: BigDecimal?,
    var basisBrutto: BigDecimal?,
    var mwstProzent: BigDecimal?,
    var hatAlternativen: Boolean,
) {
    fun isAbgelaufen(): Boolean = abgelaufen
    fun isHatAlternativen(): Boolean = hatAlternativen

    class FreigabeAnsichtDtoBuilder {
        private var uuid: String? = null
        private var status: String? = null
        private var dokumentNummer: String? = null
        private var dokumentArt: String? = null
        private var dokumentBetrag: BigDecimal? = null
        private var bauvorhaben: String? = null
        private var kundeName: String? = null
        private var kundeEmail: String? = null
        private var erstelltAm: LocalDateTime? = null
        private var ablaufDatum: LocalDateTime? = null
        private var akzeptiertAm: LocalDateTime? = null
        private var abgelaufen: Boolean = false
        private var pdfPfad: String? = null
        private var positionen: List<FreigabePositionDto>? = null
        private var basisNetto: BigDecimal? = null
        private var basisBrutto: BigDecimal? = null
        private var mwstProzent: BigDecimal? = null
        private var hatAlternativen: Boolean = false

        fun uuid(uuid: String?) = apply { this.uuid = uuid }
        fun status(status: String?) = apply { this.status = status }
        fun dokumentNummer(dokumentNummer: String?) = apply { this.dokumentNummer = dokumentNummer }
        fun dokumentArt(dokumentArt: String?) = apply { this.dokumentArt = dokumentArt }
        fun dokumentBetrag(dokumentBetrag: BigDecimal?) = apply { this.dokumentBetrag = dokumentBetrag }
        fun bauvorhaben(bauvorhaben: String?) = apply { this.bauvorhaben = bauvorhaben }
        fun kundeName(kundeName: String?) = apply { this.kundeName = kundeName }
        fun kundeEmail(kundeEmail: String?) = apply { this.kundeEmail = kundeEmail }
        fun erstelltAm(erstelltAm: LocalDateTime?) = apply { this.erstelltAm = erstelltAm }
        fun ablaufDatum(ablaufDatum: LocalDateTime?) = apply { this.ablaufDatum = ablaufDatum }
        fun akzeptiertAm(akzeptiertAm: LocalDateTime?) = apply { this.akzeptiertAm = akzeptiertAm }
        fun abgelaufen(abgelaufen: Boolean) = apply { this.abgelaufen = abgelaufen }
        fun pdfPfad(pdfPfad: String?) = apply { this.pdfPfad = pdfPfad }
        fun positionen(positionen: List<FreigabePositionDto>?) = apply { this.positionen = positionen }
        fun basisNetto(basisNetto: BigDecimal?) = apply { this.basisNetto = basisNetto }
        fun basisBrutto(basisBrutto: BigDecimal?) = apply { this.basisBrutto = basisBrutto }
        fun mwstProzent(mwstProzent: BigDecimal?) = apply { this.mwstProzent = mwstProzent }
        fun hatAlternativen(hatAlternativen: Boolean) = apply { this.hatAlternativen = hatAlternativen }

        fun build(): FreigabeAnsichtDto =
            FreigabeAnsichtDto(
                uuid,
                status,
                dokumentNummer,
                dokumentArt,
                dokumentBetrag,
                bauvorhaben,
                kundeName,
                kundeEmail,
                erstelltAm,
                ablaufDatum,
                akzeptiertAm,
                abgelaufen,
                pdfPfad,
                positionen,
                basisNetto,
                basisBrutto,
                mwstProzent,
                hatAlternativen,
            )
    }

    companion object {
        @JvmStatic
        fun builder(): FreigabeAnsichtDtoBuilder = FreigabeAnsichtDtoBuilder()
    }
}
